src/freenet/node/OpennetManager.java

```diff
@@ -546,6 +589,39 @@ public boolean wantPeer(OpennetPeerNode nodeToAddNow, boolean addAtLRU, boolean
 				return false;
 			}
 		}
+		// Don't connect to peers in excluded countries
+		if (nodeToAddNow != null) {
+			Peer[] handshakeIPs = nodeToAddNow.getHandshakeIPs();
+			if (handshakeIPs != null) {
+				for (Peer peer : handshakeIPs) {
+					if (peer == null) {
+						continue;
+					}
+					FreenetInetAddress address = peer.getFreenetAddress();
+					if (address == null) {
+						continue;
+					}
+					InetAddress inetAddress = address.getAddress(false);
+
+					IPConverter ipConverter = IPConverter.getInstance(NodeFile.IPv4ToCountry.getFile(node));
+					try {
+						Country country = ipConverter.locateIP(inetAddress.getAddress());
+						if (country != null) {
+							if (excludedCountries.contains(country.toString().toUpperCase())) {
```

Comment:
Suggested thread safety fix:
Make the excludedCountries reference volatile, and do writes to the HashSet by replacing the whole set with a new one, i.e. making the reference point to a new HashSet object.

That should be the most efficient concurrency tool for our situation of very many reads vs very few writes: The only low level effect on the reads should be that the excludedCountries variable has to be fetched from main memory every time instead of being cached, locks won't need to be acquired. The cost of creating a new HashSet upon every write can be ignored for that benefit of lock-freeness as users will rarely change the config.